### ISSUE TITLE
feat(p2p): check for a peer's reputation before returning any data

### ISSUE THREAD

**Issue by ramfox** [qri-io/qri#578]
qri-io/registry#15 and qri-io/registry_redis#16 must be merged before the ci will allow this issue to go through

## summary:

Before you exchange data with a peer, check their reputation by pinging the registry. If they have a negative reputation, close the connection.

In order to achieve this, the p2p layer will need to know the where to call to the registry, which means the registry config will also need to be surfaced in the QriNode.

#### add registryConfig to QriNode

I made the choice to add `registryConfig *config.Registry` to the QriNode struct, and I __did not__ change the `cfg` name (which houses the `*config.P2P` struct) to `p2pConfig`. If we think I should, I am happy to change it. 

let's also add a helper function for the RegistryLocation, that returns the registry location uri:
```
func (n *QriNode) RegistryLocation() string
```

`NewQriNode` should be modified to take a `*config.Config`, rather than `*config.P2P`. Ensure that anything that uses `NewQriNode` is also properly modified.

#### adding reputation to our processes:

After doing a deep dive on p2p, we learned a few things.

First, the ConnManager prioritizes connections based on the Tag values. We knew this, and this is why we tag connections with Qri peers higher than normal connections. However, and this makes so much sense in retrospect, that when a connection is closed we loose the ConnManager's tag for that connection/peer. 

So if we ever check the reputation of a peer, tag the peer as a bad actor, and then close the connection, we dont have any persistant knowledge of that we should not communicate with this peer. This means any time we create a new connection with a peer, we have to call to the registry server, to find the reputation, even if we have already done so and closed the connection previously.

Instead, let's store the reputation of a peer in the Peerstore, which allows peer related storage based on key value pairs (`peerstore.Put(peerid, keyString, value )`, and won't remove a peer from a store when a connection is closed.

Let's store the reputation with a key called `qriReputationKey`

Let's change the variable `qriConnManagerValue` to `qriBaselineReputation`. 

Let's create a new function 
`func (n *QriNode) TagReputation(pid peer.ID) error` 
- look up the reputation in the peerstore
- if it does not exists, add the baseline reputation
- if it does exist, save to a variable
- Get tag from ConnManager
- If it is the incorrect reputation:
- Untag the connection and Tag it with the reputation.

`func  (n *QriNode) UpdateReputation(pid peer.ID, reputation int) error`
- get's the reputation from the peerstore
- if one does not exist, old reputation == baseline rep
- adds the new reputation to the old reputation
- puts the reputation into the peerstore
- untags the ConnManager, and Tags with the new reputation

In `HandlePeerFound`, when we find out that a peer is indeed a Qri peer, `TagReputation`.

In AddQriPeer, once we have the profile, `checkReputation(profile *profile.Profile)`

In check Reputation, get the reputation from the registry. If no response, do nothing.

func  (n *QriNode) checkReputation(profile *profile.Profile) error
- for each peerID (found in profile.PeerIDs):
- check peerstore for reputation
- if reputation does not equal baseline reputation and is negative, call `DisconnectFromPeer` and return (since we are going to require registries to provide a reputation that is positive or negative, we know that if we have checked the reputation before, we will not get the baseline reputation back. This will prevent unnecessary calls to the registry)
- get or create regClient
- call GetReputation(profile.ID)
- if no response, return
- get the reputation from the response
- call `updateReputation(reputation)`
- get reputation from peerstore, if negative => `DisconnectFromPeer` and return

`func (n *QriNode) DisconnectFromPeer(pid peer.ID)`
- Get all connections to peer
- for each connection, close it!

Phew! very exciting

### remove unnecessary tagging and adding of Addrs.

Currently we see this in `SendMessage`, but I've noticed this in a few other places as well. We should remove these as we see them.

__last important thing to not forget: add to registry notation that we NEEEED the registry to respond with non zero numbers in order for registry to do anything effective!__

## Edit:
After doing a pass on this, we realized that we are missing something.

We need a `ProfileStore` that is an ephemeral store of a `repo.profile`. Also, `repo.Profile` needs to store a Reputation, that does not travel with the `Profile` once you convert to a `repo.ProfilePod`

To better plan for the future, we want to add a TTL to the registry reputation response. This will get discarded for now.

We need to create a `ProfileStore`:

```
type ProfileStore struct {
    profiles map[repo.ID]*repo.Profile
    peers map[peer.ID]*repo.Profile
    profilesLock sync.Mutex
    peersLock sync.Mutex
}

// Profile must have a profile.ID
func (ps *ProfileStore) SetProfile(p *repo.Profile, ttl time.Duration) error

func (ps *ProfileStore) GetProfile(id *repo.ID) *repo.Profile

func (ps *ProfileStore) GetProfileFromPeerID(id peer.ID) *repo.Profile

var ErrNotFound = fmt.Error("item not found")
```
